nomad/volumewatcher/batcher.go

```diff
@@ -67,30 +86,62 @@ func (b *VolumeUpdateBatcher) batcher() {
 			return
 		case w := <-b.workCh:
 			if timerCh == nil {
-				timerCh = time.After(b.batch)
+				timerCh = time.After(b.batchDuration)
 			}
 
+			future := NewBatchFuture()
+
+		NEXT_CLAIM:
 			// de-dupe and store the claim update, and attach the future
 			for _, upd := range w.claims {
-				claims[upd.VolumeID+upd.RequestNamespace()] = upd
+				id := upd.VolumeID + upd.RequestNamespace()
+
+				for _, batch := range batches {
+					// first see if we can dedupe the update
+					_, ok := batch.claims[id]
+					if ok {
+						batch.claims[id] = upd
+						future = batch.future
+						continue NEXT_CLAIM
+					}
+					// otherwise append to the first non-full batch
+					if len(batch.claims) < maxBatchSize {
+						batch.claims[id] = upd
+						future = batch.future
+						continue NEXT_CLAIM
+					}
+				}
+				// all batches were full, so add a new batch
+				newBatch := &claimBatch{
+					claims: map[string]structs.CSIVolumeClaimRequest{id: upd},
+					future: NewBatchFuture(),
+				}
+				batches = append(batches, newBatch)
+				future = newBatch.future
 			}
+
+			// we send batches to raft FIFO, so we return the last
+			// future to the caller so that it can wait until the
+			// last batch has been sent
 			w.f <- future
+
 		case <-timerCh:
 			// Capture the future and create a new one
-			f := future
-			future = NewBatchFuture()
 
-			// Create the batch request
+			batch := batches[0]
+			f := batch.future
+
+			// Create the batch request for the oldest batch
 			req := structs.CSIVolumeClaimBatchRequest{}
-			for _, claim := range claims {
+			for _, claim := range batch.claims {
 				req.Claims = append(req.Claims, claim)
 			}
 
 			// Upsert the claims in a go routine
 			go f.Set(b.raft.UpsertVolumeClaims(&req))
 
-			// Reset the claims list and timer
-			claims = make(map[string]structs.CSIVolumeClaimRequest)
+			// Reset the batches list and timer
+			batches = batches[1:]
```

Comment:
if i'm reading this correctly, it looks like the loop relies on work coming in on `b.workCh` to reset the timer, same as before. however, with batching, it seems like it's possible that enough work could come in during a single `batchDuration` to create multiple batches. if that happens, the `timerCh` will signal a claim request for only the first batch of claims. the rest will not be dispatched until more work comes in to reset the timer (which could hypothetically be never).

so either this loop needs to send off all batches or it needs to reset `timerCh` (to `batchDuration` or perhaps some appropriate fraction thereof) if `len(batches) > 0`.

Reply:
aside: while i appreciate the elegance of this approach, why just not use a `Ticker` for `timerCh`?

Reply:
> so either this loop needs to send off all batches or it needs to reset timerCh (to batchDuration or perhaps some appropriate fraction thereof) if len(batches) > 0.

Good catch, that wasn't a problem when we were sending off the whole batch at once!

> aside: while i appreciate the elegance of this approach, why just not use a Ticker for timerCh?

Hm, that's a good point. In the original design (copied right out of [`deploymentwatcher`](https://github.com/hashicorp/nomad/blob/master/nomad/deploymentwatcher/batcher.go)), a no-op pass would potentially modify some state by swapping out the future. A ticker would make it a little easier to understand at the negligible cost of ticking over and checking the length each pass.

Reply:
I've updated with these changes. How's that look?